### PR TITLE
Buf speedup

### PR DESCRIPTION
**Pull request by jcorbin**
From a preliminary test, this proves to be 30% and 50% faster for trivial strings (random sampling from /usr/share/dict/words).

### CODE REVIEW

**Comment by Raynos** [comment]
This looks ok. I would recommend testing a bunch of strings, maybe use these ones ( https://github.com/joyent/node/blob/master/test/simple/test-string-decoder.js#L28-L53 ).

File: lib/utf8len.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return 5;
+    } else if (c < 0x80000000) {
+        // 1111110b 10bbbbbb 10bbbbbb 10bbbbbb 10bbbbbb 10bbbbbb
+        return 6;
+    } else {
+        throw new Error('codepoint out of range');
+    }
+}
+
+function numBytesForUtf8String(s) {
```
Comment (Raynos) [inline lib/utf8len.js]: lets leave a comment about this being faster then `Buffer.byteLength`

**Comment by jcorbin** [comment]
Interestingly using my utf8len function seems to be 10% - 35% faster than using the `Buffer.byteLength(source, 'utf-8')`

**Comment by jcorbin** [comment]
afaik, correctness of utf8len is down to correctness of `String.prototype.charCodeAt`

**Comment by kriskowal** [comment]
@jcorbin bear in mind that String.prototype.charCodeAt is ucs2, not unicode in general. If folks express code points outside the Basic Multilingual Plane, they have to resort to surrogate pairs. Not sure whether utf8len takes that into account, but we may be able to safely punt the issue. https://mathiasbynens.be/notes/javascript-unicode

**Comment by jcorbin** [comment]
v2 about to dead land, will reboot if needed

**Pull request closed without merging**